src/main/java/com/celements/filebase/uri/FileUriService.java

```diff
@@ -0,0 +1,160 @@
+  private IAttachmentServiceRole attachmentSrv;
+
+  @Requirement
+  private LastStartupTimeStampRole lastStartupTimeStamp;
+
+  @Requirement(CelementsFromWikiConfigurationSource.NAME)
+  private ConfigurationSource configSrc;
+
+  private String getDefaultAction() {
+    return Optional.ofNullable(
+        Strings.emptyToNull(configSrc.getProperty("celements.fileuri.defaultaction")))
+        .orElse(context.getXWikiContext().getWiki().getXWikiPreference("celdefaultAttAction",
+            "celements.attachmenturl.defaultaction", "file", context.getXWikiContext()));
+  }
+
+  @Override
+  public @NotNull UriBuilder getFileUriPrefix(@NotNull Optional<String> action) {
+    URL baseUrl = getUrlFactory().createResourceURL("", false, context.getXWikiContext());
+    try {
+      return UriBuilder.fromUri(baseUrl.toURI())
+          .replacePath("/" + action.orElse(getDefaultAction()) + "/")
+          .path(baseUrl.getPath());
+    } catch (URISyntaxException exp) {
+      LOGGER.error("Failed to get file url prefix.", exp);
+      return UriBuilder.fromPath(baseUrl.toString());
+    }
+  }
+
+  @Override
+  @NotNull
+  public UriBuilder createAbsoluteFileUri(@NotNull FileReference fileRef,
+      @NotNull Optional<String> action, Optional<String> queryString) {
+    try {
+      return UriBuilder.fromUri(getUrlFactory().getServerURL(context.getXWikiContext()).toURI())
+          .uri(createFileUri(fileRef, action, queryString).build());
+    } catch (MalformedURLException | FileNotExistException | URISyntaxException exp) {
+      LOGGER.error("Failed to getServerURL for [{}].", fileRef, exp);
+      return UriBuilder.fromPath("");
+    }
+  }
+
+  private XWikiURLFactory getUrlFactory() {
+    return context.getXWikiContext().getURLFactory();
+  }
+
+  @Override
+  @NotNull
+  public UriBuilder createFileUri(@NotNull FileReference fileRef, @NotNull Optional<String> action,
+      @NotNull Optional<String> queryString) throws FileNotExistException {
+    final UriBuilder baseUrl = createFileUri(fileRef, action);
+    if (queryString.isPresent()) {
+      return baseUrl.replaceQuery(Optional.ofNullable(baseUrl.build().getQuery())
+          .map(qS -> qS + "&" + queryString.get())
+          .orElse(queryString.get()));
+    }
+    return baseUrl;
+  }
+
+  @NotNull
+  UriBuilder createFileUri(@NotNull FileReference fileRef,
+      @NotNull Optional<String> action)
+      throws FileNotExistException {
+    UriBuilder uriBuilder;
+    if (fileRef.isAttachmentReference()) {
+      uriBuilder = createAttachmentUri(fileRef, action);
+    } else if (fileRef.isOnDiskReference()) {
+      uriBuilder = createOnDiskUri(fileRef, action);
+    } else {
+      uriBuilder = fileRef.getUri();
+    }
+    return addContextUri(uriBuilder);
+  }
+
+  UriBuilder addContextUri(UriBuilder uriBuilder) {
+    Optional<XWikiDocument> currentDoc = context.getCurrentDoc().toJavaUtil();
+    if (currentDoc.isPresent() && uriBuilder.toString().startsWith("?")) {

```

Comment:
I don't think checking the toString of the builder is a good idea since we have no guarantee as to how it's implemented apart from being a "textual" representation. perhaps you want to check for no path?

Reply:
Well, there is no other way to check the state of the URI-Builder. Furthermore "toString" is the only way to NOT get a "IlligaleArgumentException" for not fully qualified urls.

Reply:
I see, it's really unfortunate that the UriBuilder interface doesn't have getters for the URI parts.